src/node_api.cc

```diff
@@ -23,12 +23,16 @@ void napi_clear_last_error(napi_env env);
 
 class napi_env__ {
  public:
-  explicit napi_env__(v8::Isolate* _isolate): isolate(_isolate), last_error() {}
+  explicit napi_env__(v8::Isolate* _isolate): isolate(_isolate),
+      has_instance_available(true), last_error() {}
   ~napi_env__() {
     last_exception.Reset();
+    has_instance.Reset();
   }
   v8::Isolate* isolate;
   v8::Persistent<v8::Value> last_exception;
+  v8::Persistent<v8::Value> has_instance;
```

Comment:
I just wondered if the exact same code, except that having has_instance be static would work.  The cache would be created the first time an env was created.  The only concern was whether this could happen in parallel, but given Node.js generally single thread nature that may not be an issue.